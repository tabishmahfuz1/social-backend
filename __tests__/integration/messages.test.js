const { gql } = require("apollo-server");
const faker = require("faker");
const { testClient } = require("../util/testClient");
const typeDefs = require("../../src/graphql/typeDefs");
const resolvers = require("../../src/graphql/resolvers/index");
const User = require("../../src/models/User");
const generateToken = require("../../src/util/generateToken");

const CREATE_MESSAGE = gql`
  mutation createMessage($notifier: String!, $body: String!, $threadId: ID!) {
    createMessage(notifier: $notifier, body: $body, threadId: $threadId) {
      creator {
        firstName
        lastName
        avatarImage
      }
      notifier {
        firstName
        lastName
        avatarImage
      }
      body
    }
  }
`;

const CREATE_THREAD = gql`
  mutation createThread($urlUser: String!) {
    createThread(urlUser: $urlUser) {
      id
      participantsIds
      createdAt
    }
  }
`;

const GET_CONVERSATIONS = gql`
  {
    getConversations {
      latestMessage {
        creator {
          firstName
          lastName
          avatarImage
        }
        notifier {
          firstName
          lastName
          avatarImage
        }
        body
      }
    }
  }
`;

describe("Chat system integration testing", () => {
  test("User A and User B: User A should be able to message User B", async () => {
    // ARRANGE
    const userA = await new User({
      firstName: "John",
      lastName: "Doe",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const userB = await new User({
      firstName: "Jack",
      lastName: "Daniels",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const token = generateToken(userA);

    // ACT
    // use the test server to create a query function
    const { query, mutate } = testClient(
      {
        resolvers,
        typeDefs,
        context: async ({ req }) => ({ req }),
      },
      {
        req: { headers: { authorization: `JWT ${token}` } },
      }
    );

    // create a thread that is going to be used for all messages between user A and user B ( conversation ) from authenticated user ( user A )
    const thread = await mutate({
      mutation: CREATE_THREAD,
      variables: {
        urlUser: userB.id,
      },
    });

    // create a message from authenticated user ( user A )
    const message = await mutate({
      mutation: CREATE_MESSAGE,
      variables: {
        notifier: userB.id,
        body: "how are you buddy?",
        threadId: thread.data.createThread.id,
      },
    });
    // assert that createMessage mutation returns the proper data
    expect(message).toMatchObject({
      data: {
        createMessage: {
          creator: {
            firstName: userA.firstName,
            lastName: userA.lastName,
          },
          notifier: {
            firstName: userB.firstName,
            lastName: userB.lastName,
          },
          body: "how are you buddy?",
        },
      },
    });

    // get messages for the notifier ( in this case authenticated user: User B) and assert that there is a new message from user B
    const tokenB = generateToken(userB);

    const conversations = await query({
      mutation: GET_CONVERSATIONS,
      ctxArg: { req: { headers: { authorization: `JWT ${tokenB}` } } },
    });
    expect(conversations).toMatchSnapshot();
  });

  // Subscriptions cannot be tested as apollo-server-testing is not allowing it yet.
});
