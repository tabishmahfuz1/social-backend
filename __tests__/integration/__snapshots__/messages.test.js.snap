// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat system integration testing User A and User B: User A should be able to message User B 1`] = `
Object {
  "data": Object {
    "getConversations": Array [
      Object {
        "latestMessage": Object {
          "body": "how are you buddy?",
          "creator": Object {
            "avatarImage": "https://scontent.fsof4-1.fna.fbcdn.net/v/t1.30497-1/c71.0.240.240a/p240x240/84241059_189132118950875_4138507100605120512_n.jpg?_nc_cat=1&_nc_sid=7206a8&_nc_ohc=CD3ImewvDlgAX8f-QvY&_nc_ht=scontent.fsof4-1.fna&oh=782172992e9272da49f35f7f40ef8d63&oe=5EED0201",
            "firstName": "John",
            "lastName": "Doe",
          },
          "notifier": Object {
            "avatarImage": "https://scontent.fsof4-1.fna.fbcdn.net/v/t1.30497-1/c71.0.240.240a/p240x240/84241059_189132118950875_4138507100605120512_n.jpg?_nc_cat=1&_nc_sid=7206a8&_nc_ohc=CD3ImewvDlgAX8f-QvY&_nc_ht=scontent.fsof4-1.fna&oh=782172992e9272da49f35f7f40ef8d63&oe=5EED0201",
            "firstName": "Jack",
            "lastName": "Daniels",
          },
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
